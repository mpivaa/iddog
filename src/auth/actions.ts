import { Dispatch } from 'redux';
import { post } from '@app/api';
import { push } from '@app/router';
import { Error } from '@app/types';

export enum ActionType {
  REQUEST = 'AUTH_REQUEST',
  SUCCESS = 'AUTH_SUCCESS',
  FAILURE = 'AUTH_FAILURE',
}

export const authRequest = () => ({ type: ActionType.REQUEST });
export const authSuccess = (token: string) => ({ type: ActionType.SUCCESS, token });
export const authError = (error: Error) => ({ type: ActionType.FAILURE, error });

interface SignUpPayload {
  user: {
    token: string,
  };
}

export function signUp(email: string) {
  return async (dispatch: Dispatch) => {
    dispatch(authRequest());
    const { data, error } = await post<SignUpPayload>('/signup', {
      params: {
        email,
      },
    });

    if (error) {
      return dispatch(authError(error));
    } else {
      dispatch(authSuccess(data!.user.token));
      return dispatch(push('/feed'));
    }
  };
}

export type Action =
  | { type: typeof ActionType.REQUEST }
  | { type: typeof ActionType.SUCCESS, token: string }
  | { type: typeof ActionType.FAILURE, error: Error };
