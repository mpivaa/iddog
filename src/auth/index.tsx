export { default as SignUp } from './SignUp';
export { default as reducer, State } from './reducer';
export * from './selectors';
export * from './actions';
