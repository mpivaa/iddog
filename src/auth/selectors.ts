import { State } from '@app/store';

export const getToken = (state: State) => state.auth.token;
export const getError = (state: State) => state.auth.error;
export const getLoading = (state: State) => state.auth.loading;
