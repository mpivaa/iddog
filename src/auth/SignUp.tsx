import * as React from 'react';
import { connect } from 'react-redux';
import { signUp } from '@app/auth/actions';
import { getError, getLoading } from '@app/auth/selectors';
import { State as AppState } from '@app/store';
import { Error as ErrorType } from '@app/types';
import { Button, Container, Error, Input, LinearLayout, Title } from '@app/ui';

interface State {
  readonly email: string;
}

interface Props {
  error: ErrorType | null;
  loading: boolean;
  signUp(token: string): void;
}

export class SignUp extends React.Component<Props, State> {
  readonly state: State = { email: '' };

  signUp(e: React.SyntheticEvent) {
    e.preventDefault();
    this.props.signUp(this.state.email);
    return false;
  }

  render() {
    const { error, loading } = this.props;

    return (
      <Container size="small">
        <LinearLayout direction="column" fill>
          <form onSubmit={(e) => this.signUp(e)}>
            <Title>The IDDOG</Title>
            <Input
              placeholder="your email"
              type="email"
              value={this.state.email}
              onChange={(e) => this.setState({ email: e.target.value })}
            />
            <Error>{error && error.message}</Error>
            <Button loading={loading}>
              Entrar
            </Button>
          </form>
        </LinearLayout>
      </Container>
    );
  }
}

const mapState = (state: AppState) => ({
  error: getError(state),
  loading: getLoading(state),
});

const mapDispatch = {
  signUp,
};

export default connect(mapState, mapDispatch)(SignUp);
