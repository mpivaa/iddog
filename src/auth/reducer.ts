import { Action, ActionType } from '@app/auth/actions';
import { Error } from '@app/types';

export interface State {
  loading: boolean;
  token: string | null;
  error: Error | null;
}

const initialState: State = {
  loading: false,
  token: null,
  error: null,
};

function reducer(state = initialState, action: Action) {
  switch (action.type) {
    case ActionType.REQUEST: return {
      ...state,
      loading: true,
    };
    case ActionType.SUCCESS: return {
      ...state,
      loading: false,
      token: action.token,
      error: null,
    };
    case ActionType.FAILURE: return {
      ...state,
      loading: false,
      token: null,
      error: action.error,
    };
    default: return state;
  }
}

export default reducer;
