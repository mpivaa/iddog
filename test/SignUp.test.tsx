import * as React from 'react';
import { mount } from 'enzyme';
import { SignUp } from '@app/auth';
import Setup, { store, fetchMock } from '@test/setup';

test('SignUp renders correctly', () => {
  const component = mount(
    <Setup>
      <SignUp />
    </Setup>
  );

  expect(component.render()).toMatchSnapshot();
});

test('Shows an error when email is not valid', (done) => {
  fetchMock.once('end:signup', {
    status: 404,
    body: { error: { message: 'Some error' }}
  });

  const component = mount(
    <Setup>
      <SignUp />
    </Setup>
  );

  component.find('input').simulate('change', {
    target: { value: 'invalid-email' }
  });

  component.find('form').simulate('submit');

  setTimeout(() => {
    expect(store.getState()).toMatchSnapshot();
    expect(component.render()).toMatchSnapshot();
    fetchMock.restore();
    done();
  }, 100);
});

test('Sets token and show feed when authorized', (done) => {
  fetchMock.once('end:signup', {
    user: { token: 'xyz' }
  });

  const component = mount(
    <Setup>
      <SignUp />
    </Setup>
  );

  component.find('input').simulate('change', {
    target: { value: 'valid@email.com' }
  });

  component.find('form').simulate('submit');

  setTimeout(() => {
    expect(store.getState().auth).toMatchSnapshot();
    expect(store.getState().router.location.pathname).toBe('/feed');
    fetchMock.restore();
    done();
  }, 100);
});
