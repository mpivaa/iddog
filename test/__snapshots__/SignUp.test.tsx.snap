// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sets token and show feed when authorized 1`] = `
Object {
  "error": null,
  "loading": false,
  "token": "xyz",
}
`;

exports[`Shows an error when email is not valid 1`] = `
Object {
  "auth": Object {
    "error": Object {
      "message": "Some error",
    },
    "loading": false,
    "token": null,
  },
  "feed": Object {
    "data": Object {
      "category": "",
      "list": Array [],
    },
    "error": null,
    "loading": false,
  },
  "router": Object {
    "action": "POP",
    "location": Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    },
  },
}
`;

exports[`Shows an error when email is not valid 2`] = `
<div
  class="sc-bxivhb dHxAwy"
>
  <div
    class="sc-htpNat dSBIma"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <input
      class="sc-bdVaJa eKhXFQ"
      placeholder="your email"
      type="email"
      value="invalid-email"
    />
    <div
      class="sc-ifAKCX iIcXGD"
    >
      Some error
    </div>
    <button
      class="sc-bwzfXH emADTk"
    >
      Entrar
    </button>
  </div>
</div>
`;

exports[`SignUp renders correctly 1`] = `
<div
  class="sc-bxivhb dHxAwy"
>
  <div
    class="sc-htpNat dSBIma"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <input
      class="sc-bdVaJa eKhXFQ"
      placeholder="your email"
      type="email"
      value=""
    />
    <div
      class="sc-ifAKCX iIcXGD"
    />
    <button
      class="sc-bwzfXH emADTk"
    >
      Entrar
    </button>
  </div>
</div>
`;
